# <center>Iris species prediction using KNN and LogisticRegression classification</center><hr>

In [1]:
from sklearn.datasets import load_iris

In [7]:
iris = load_iris()
print(type(iris))

<class 'sklearn.utils.Bunch'>


In [8]:
print(iris.data)

[[5.1 3.5 1.4 0.2]
 [4.9 3.  1.4 0.2]
 [4.7 3.2 1.3 0.2]
 [4.6 3.1 1.5 0.2]
 [5.  3.6 1.4 0.2]
 [5.4 3.9 1.7 0.4]
 [4.6 3.4 1.4 0.3]
 [5.  3.4 1.5 0.2]
 [4.4 2.9 1.4 0.2]
 [4.9 3.1 1.5 0.1]
 [5.4 3.7 1.5 0.2]
 [4.8 3.4 1.6 0.2]
 [4.8 3.  1.4 0.1]
 [4.3 3.  1.1 0.1]
 [5.8 4.  1.2 0.2]
 [5.7 4.4 1.5 0.4]
 [5.4 3.9 1.3 0.4]
 [5.1 3.5 1.4 0.3]
 [5.7 3.8 1.7 0.3]
 [5.1 3.8 1.5 0.3]
 [5.4 3.4 1.7 0.2]
 [5.1 3.7 1.5 0.4]
 [4.6 3.6 1.  0.2]
 [5.1 3.3 1.7 0.5]
 [4.8 3.4 1.9 0.2]
 [5.  3.  1.6 0.2]
 [5.  3.4 1.6 0.4]
 [5.2 3.5 1.5 0.2]
 [5.2 3.4 1.4 0.2]
 [4.7 3.2 1.6 0.2]
 [4.8 3.1 1.6 0.2]
 [5.4 3.4 1.5 0.4]
 [5.2 4.1 1.5 0.1]
 [5.5 4.2 1.4 0.2]
 [4.9 3.1 1.5 0.2]
 [5.  3.2 1.2 0.2]
 [5.5 3.5 1.3 0.2]
 [4.9 3.6 1.4 0.1]
 [4.4 3.  1.3 0.2]
 [5.1 3.4 1.5 0.2]
 [5.  3.5 1.3 0.3]
 [4.5 2.3 1.3 0.3]
 [4.4 3.2 1.3 0.2]
 [5.  3.5 1.6 0.6]
 [5.1 3.8 1.9 0.4]
 [4.8 3.  1.4 0.3]
 [5.1 3.8 1.6 0.2]
 [4.6 3.2 1.4 0.2]
 [5.3 3.7 1.5 0.2]
 [5.  3.3 1.4 0.2]
 [7.  3.2 4.7 1.4]
 [6.4 3.2 4.5 1.5]
 [6.9 3.1 4.

In [10]:
# print coloumn names / attributes
print(iris.feature_names)

['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)', 'petal width (cm)']


In [12]:
# print integers representing species of each observation/record/row
# target represents what we're going to predict = dependent variable/outcome/response/label

print(iris.target)

[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 2 2 2 2 2 2 2 2 2 2
 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
 2 2]


In [14]:
# 0 = setosa
# 1 = versicolor
# 2 = virginica
print(iris.target_names)

['setosa' 'versicolor' 'virginica']


### _Two types of supervised learning_
#### _1. classification = dependent var is categorical = values are finite,unordered set of values_
#### _2. regression = where dependent var is ordered and continuous_
#### Here we've categorical values and we're using "classification technique" here

### _req of scikit-learn_
#### _1. features and response are seperate objs_
#### _2. features and response are numbers_
#### _3. features and response are numpy arrays_
#### _4. features and response have specific shapes_

In [15]:
#check type of feature, response
print(type(iris.data))
print(type(iris.target))

<class 'numpy.ndarray'>
<class 'numpy.ndarray'>


In [19]:
print(iris.data.shape)

(150, 4)


In [17]:
print(iris.target.shape)

(150,)


In [22]:
# capital X for storing feature matrix
X = iris.data

#small y for storing response vector
y = iris.target

### _When predicting the species, we'll see the measurement of unknown and see for its similiar measurements (nearest neighbours) and predict that species for unknown._
### _This is similiar to KNN classification (K nearest neighbour) _

###  <hr>KNN steps:
#### 1. pick value for k
#### 2. search for k records in training data that are nearest to unknown iris
#### 3. use most popular response/target value i.e. species name from k nearest neighbours for unknown iris<hr>

## *Model calculates numerical dist between unknown iris and each 150 known iris*
## *Then selects k known irises with smallest dist to unknown iris*
## _Euclidean dist is often used_

In [24]:
print(X.shape)
print(y.shape)

(150, 4)
(150,)


### <hr>scikit-learn 4 step modelling pattern
#### _step1:import class_

In [30]:
from sklearn.neighbors import KNeighborsClassifier

#### _step2: instantiate the estimator (make instance of model)_

In [28]:
knn = KNeighborsClassifier(n_neighbors =1) #hyperparameter initialized
#other values r set to  default
knn

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
                     metric_params=None, n_jobs=None, n_neighbors=1, p=2,
                     weights='uniform')

#### _step3: training the model - fit the model with data"_

In [29]:
# model is learning relationships betwn X and y
# traning occurs in-place hence we don't need to assign result to other obj
knn.fit(X,y)

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
                     metric_params=None, n_jobs=None, n_neighbors=1, p=2,
                     weights='uniform')

#### _step4: predict the response for new obj_

In [32]:
# new observations= out-of-sample data
# giving measurements of unknowwn species and ask to predict based on given training

knn.predict([[3, 5, 4, 2]])

array([2])

#### _it predicted as it belongs to species at index 2 in target i.e.virginica_

In [38]:
knn.predict([[3, 5, 4, 2],[5, 4, 3, 2]]) #giving 2 unknown sample measurements

array([2, 1])

## <hr>Use diff values for K
### _varying the arguments that you've passed to the model is called as model tuning_

In [41]:
#  instantiate the model using diff value
knn = KNeighborsClassifier(n_neighbors=5)

In [42]:
#fit model
knn.fit(X, y)

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
                     metric_params=None, n_jobs=None, n_neighbors=5, p=2,
                     weights='uniform')

In [44]:
# x_new = [[3, 5, 4, 2],[5, 4, 3, 2]]
knn.predict([[3, 5, 4, 2],[5, 4, 3, 2]])

array([1, 1])

## <hr>Use diff classification model
#### _we can use same 4 step pattern for deff models easily_

In [46]:
# Logistic regression is other model used for classification

# step1 import class for model
from sklearn.linear_model import LogisticRegression

In [52]:
# step2 instantiate the model
logreg = LogisticRegression(max_iter=1700) #default max_iter is 100
logreg

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=1700,
                   multi_class='auto', n_jobs=None, penalty='l2',
                   random_state=None, solver='lbfgs', tol=0.0001, verbose=0,
                   warm_start=False)

In [53]:
# step3 fit the model
logreg.fit(X, y)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=1700,
                   multi_class='auto', n_jobs=None, penalty='l2',
                   random_state=None, solver='lbfgs', tol=0.0001, verbose=0,
                   warm_start=False)

In [54]:
# step4 predict the response for unknown record
x_new = [[3, 5, 4, 2],[5, 4, 3, 2]]
logreg.predict(x_new)

array([0, 0])